File: libraries/cpp/src/writer.cpp
#include "ifcx/writer.h"

#include <fstream>
#include <stdexcept>

namespace ifcx {

std::string IfcxWriter::to_string(const IfcxDocument& doc, int indent) {
    IfcxFile file = doc.to_json();
    nlohmann::json j = file;
    return j.dump(indent);
}

void IfcxWriter::to_file(const IfcxDocument& doc,
                          const std::filesystem::path& path,
                          int indent) {
    std::ofstream ofs(path);
    if (!ofs.is_open()) {
        throw std::runtime_error("Failed to open file for writing: " + path.string());
    }

    IfcxFile file = doc.to_json();
    nlohmann::json j = file;
    ofs << j.dump(indent) << '\n';
}

} // namespace ifcx

File: libraries/python/ifcx/writer.py
"""IFCX file writer."""

from __future__ import annotations

from pathlib import Path

from ifcx.document import IfcxDocument


class IfcxWriter:
    """Writes IFCX (JSON) files."""

    @staticmethod
    def to_file(doc: IfcxDocument, path: str | Path, indent: int = 2) -> None:
        """Write IFCX to file."""
        path = Path(path)
        path.write_text(doc.to_json(indent=indent), encoding="utf-8")

    @staticmethod
    def to_string(doc: IfcxDocument, indent: int = 2) -> str:
        """Write IFCX to JSON string."""
        return doc.to_json(indent=indent)

    @staticmethod
    def to_bytes(doc: IfcxDocument, indent: int = 2) -> bytes:
        """Write IFCX to bytes."""
        return doc.to_json(indent=indent).encode("utf-8")
